src/mysql_protocol.erl

```diff
@@ -73,36 +81,65 @@ handshake(Username, Password, Database, SockModule0, SSLOpts, Socket0,
             Response = build_handshake_response(Handshake, Username, Password,
                                                 Database, SetFoundRows),
             {ok, SeqNum3} = send_packet(SockModule, Socket, Response, SeqNum2),
-            handshake_finish_or_switch_auth(Handshake, Password, SockModule,
-                                            Socket, SeqNum3);
+            handshake_finish_or_switch_auth(Handshake, Password, SockModule, Socket,
+                                            SeqNum3);
         #error{} = Error ->
             Error
     end.
 
-handshake_finish_or_switch_auth(Handshake = #handshake{status = Status}, Password,
-                                SockModule, Socket, SeqNum0) ->
-    {ok, ConfirmPacket, SeqNum1} = recv_packet(SockModule, Socket, SeqNum0),
-    case parse_handshake_confirm(ConfirmPacket) of
+
+handshake_finish_or_switch_auth(Handshake, Password, SockModule, Socket, SeqNum) ->
+    #handshake{auth_plugin_name = AuthPluginName,
+	       auth_plugin_data = AuthPluginData,
+	       server_version = ServerVersion,
+	       status = Status} = Handshake,
+    AuthResult = auth_finish_or_switch(AuthPluginName, AuthPluginData, Password,
+				       SockModule, Socket, ServerVersion, SeqNum),
+    case AuthResult of
         #ok{status = OkStatus} ->
             %% check status, ignoring bit 16#4000, SERVER_SESSION_STATE_CHANGED
             %% and bit 16#0002, SERVER_STATUS_AUTOCOMMIT.
             BitMask = bnot (?SERVER_SESSION_STATE_CHANGED bor ?SERVER_STATUS_AUTOCOMMIT),
             StatusMasked = Status band BitMask,
             StatusMasked = OkStatus band BitMask,
             {ok, Handshake, SockModule, Socket};
-        #auth_method_switch{auth_plugin_name = AuthPluginName,
-                            auth_plugin_data = AuthPluginData} ->
-            Hash = case AuthPluginName of
-                       <<>> ->
-                           hash_password(Password, AuthPluginData);
-                       <<"mysql_native_password">> ->
-                           hash_password(Password, AuthPluginData);
-                       UnknownAuthMethod ->
-                           error({auth_method, UnknownAuthMethod})
-                   end,
+	Error ->
+            Error
+    end.
+
+auth_finish_or_switch(AuthPluginName, AuthPluginData, Password,
+                      SockModule, Socket, ServerVersion, SeqNum0) ->
+    {ok, ConfirmPacket, SeqNum1} = recv_packet(SockModule, Socket, SeqNum0),
+    case parse_handshake_confirm(AuthPluginName, ConfirmPacket) of
+        #ok{} = Ok ->
+            Ok;
+        #auth_method_switch{auth_plugin_name = SwitchAuthPluginName,
+                            auth_plugin_data = SwitchAuthPluginData} ->
+            Hash = hash_password(SwitchAuthPluginName, Password, SwitchAuthPluginData),
```

Comment:
And believe me, after doing this, I would prefer not to know ;)

You may look at [this flow chart](https://dev.mysql.com/doc/dev/mysql-server/latest/page_caching_sha2_authentication_exchanges.html) if you like your brain twisted, or read [this more easily digestable text](http://www.dataarchitect.cloud/preparing-your-community-connector-for-mysql-8-part-2-sha256/). The entire things is very underdocumented from MySQLs side, most of it I got from the above two sources, odd bits and pieces from Google, and by looking at the Java driver. And part is all guesswork.

Anyway, this is a very involved thing. It would require a very extensive comment. Though, I don't know where this may be better placed.

Reply:
I don't really want to get my brain twisted :-) but I want to keep the code maintainable. One sentence at the definition of each function can be good enough, just to make it easier to follow the code. At some other place, you added quite good comments. These links can be added in code comments, if useful. Guesswork/assumptions are good to document too, but just a line or so.

Reply:
Ok, I added a general comment for the entire function, and individual ones in the `case` branches. What do you think?

Reply:
Awesome